compile/src/main/scala/sbt/compiler/javac/DiagnosticsReporter.scala

```diff
@@ -69,18 +69,37 @@ final class DiagnosticsReporter(reporter: Reporter) extends DiagnosticListener[J
         def startPosition: Option[Long] = checkNoPos(d.getStartPosition)
         def endPosition: Option[Long] = checkNoPos(d.getEndPosition)
         override val offset: Maybe[Integer] = Logger.o2m(checkNoPos(d.getPosition) map { x => new Integer(x.toInt) })
-        // TODO - Is this pulling contents of the line correctly?
-        // Would be ok to just return null if this version of the JDK doesn't support grabbing
-        // source lines?
-        override def lineContent: String =
-          Option(d.getSource) match {
-            case Some(source: JavaFileObject) =>
-              (Option(source.getCharContent(true)), startPosition, endPosition) match {
-                case (Some(cc), Some(start), Some(end)) => cc.subSequence(start.toInt, end.toInt).toString
-                case _                                  => ""
+        override def lineContent: String = {
+          def getDiagnosticLine: Option[String] =
+            try {
+              // See com.sun.tools.javac.api.ClientCodeWrapper.DiagnosticSourceUnwrapper
+              val diagnostic = d.getClass.getField("d").get(d)
+              // See com.sun.tools.javac.util.JCDiagnostic#getDiagnosticSource
+              val getDiagnosticSourceMethod = diagnostic.getClass.getDeclaredMethod("getDiagnosticSource")
+              Option(getDiagnosticSourceMethod.invoke(diagnostic)) match {
+                case Some(diagnosticSource) =>
+                  // See com.sun.tools.javac.util.DiagnosticSource
+                  val getLineMethod = diagnosticSource.getClass.getMethod("getLine", Integer.TYPE)
+                  Option(getLineMethod.invoke(diagnosticSource, line.get())).map(_.toString)
+                case _ => None
               }
-            case _ => ""
-          }
+            } catch {
+              case ignored: NoSuchMethodException => None
```

Comment:
Very glad to see this caught and returning None.  I think this whole "getDiagnosticLine" mechanism should probably be its own stand-alone static utiltiy method.